### PR TITLE
Adding POM-Operator code to Codemodder repository

### PR DESCRIPTION
**Pull request by carlosu7**
Bringing [POM-OPERATOR java source code migration](https://github.com/pixee/pom-operator/pull/37) classes to Codemodder project.

Also [POM-Operator java tests from this branch](https://github.com/carlosu7/pom-operator/pull/4) were added to codemodder repo

Basically what I did was:

* Moved POM-Operator java source code to `plugins/codemodder-plugin-maven/src/main/java/io.codemodder.plugins.maven.operator` 
* Moved POM-Operator java tests to `plugins/codemodder-plugin-maven/src/test/java/io.codemodder.plugins.maven.operator`
* Moved POM-Operator xml test resources to `plugins/codemodder-plugin-maven/src/test/resources/io.codemodder.plugins.maven.operator`. 
* Update `plugins/codemodder-plugin-maven/build.gradle.kts` to add pom-operator dependencies

### CODE REVIEW

**Comment by carlosu7** [comment]
@aldrinleal when I was locally running tests for pom-operator, I was not able to run `POMOperatorJavaIT.java` successfully, but with `act -j build` it runs successfully. With that said, when I'm locally running `./gradlew check` for codemodder, `POMOperatorJavaIT` keeps failing.

File: plugins/codemodder-plugin-maven/java-sample/src/test/java/io/github/pixee/it/POMOperatorJavaTest.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import io.github.pixee.maven.operator.ProjectModel;
+import org.dom4j.DocumentException;
+import org.junit.Test;
+
+import io.github.pixee.maven.operator.Dependency;
+
+import javax.xml.stream.XMLStreamException;
+import java.io.IOException;
+import java.net.URISyntaxException;
+
```
Comment (nahsra) [inline plugins/codemodder-plugin-maven/java-sample/src/test/java/io/github/pixee/it/POMOperatorJavaTest.java:14]: We should make all the tests package-protected and `final`

File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractCommand.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
+package io.codemodder.plugins.maven.operator;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.List;
+import javax.xml.stream.XMLStreamException;
+import org.dom4j.Element;
+import org.dom4j.Node;
+
```
Comment (nahsra) [inline plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractCommand.java:10]: While you have good context here on these, can you add JavaDoc?

File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractQueryCommand.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import org.apache.commons.lang3.SystemUtils;
+import org.apache.maven.shared.invoker.DefaultInvocationRequest;
+import org.apache.maven.shared.invoker.InvocationRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (nahsra) [inline plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractQueryCommand.java:16]: JavaDoc

Does it need to be `public`?

File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractQueryCommand.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    return outputPath;
+  }
+
+  protected File getPomFilePath(POMDocument d) throws URISyntaxException {
+    Path pomPath = Paths.get(d.getPomPath().toURI());
+    return pomPath.toFile();
+  }
+
+  protected abstract void extractDependencyTree(File outputPath, File pomFilePath, ProjectModel c);
```
Comment (nahsra) [inline plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractQueryCommand.java:39]: JavaDoc

File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractVersionCommand.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package io.codemodder.plugins.maven.operator;
+
+import java.util.*;
+import org.apache.commons.lang3.builder.CompareToBuilder;
+
+public class AbstractVersionCommand extends AbstractCommand {
```
Comment (nahsra) [inline plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractVersionCommand.java:6]: JavaDoc

File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractVersionCommand.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  public Set<VersionDefinition> result = new TreeSet<>(VERSION_KIND_COMPARATOR);
```
Comment (nahsra) [inline plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractVersionCommand.java:8]: Does this need to be `public`?

File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/Chain.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
+package io.codemodder.plugins.maven.operator;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.*;
+import javax.xml.stream.XMLStreamException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Chain {
```
Comment (nahsra) [inline plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/Chain.java:10]: This class name is too generic, what's it a chain of? I think `CommandChain` makes sense.

**Review by nahsra — CHANGES_REQUESTED** [review]
Can you take at removing `public` everywhere it's not needed? Without this, it will be a little hard to work on the monorepo because of all these new classes being visible everywhere.

Whatever APIs need to be `public` after that, let's add JavaDoc to them. There's a lot of other cleanup we can do, but it can wait. My main concern is adding 100+ new public Java classes to the project.

**Comment by aldrinleal** [comment]
> @aldrinleal @nahsra I tag you here since I'm not able to add reviewers to this PR
> 
> @aldrinleal when I was locally running tests for pom-operator, I was not able to run `POMOperatorJavaIT.java` successfully, but with `act -j build` it runs successfully. With that said, when I'm locally running `./gradlew check` for codemodder, `POMOperatorJavaIT` keeps failing.
> 
> Also, I wonder if content inside`plugins/codemodder-plugin-maven/java-sample/src` is needed. I see that `plugins/codemodder-plugin-maven/java-sample/pom.xml` is needed for `POMOperatorJavaIT.java`

`POMOperatorJavaIT` was a safe-measure (or another canary, just like the `-result.xml` files) to ensure it was able to be consumed from plain Java. You can safely delete that since its on the same codebase after all.

**Review by nahsra — APPROVED** [review]

**Pull request merged**